[PyTorch Distributed] Fix batch_isend_irecv

kwen2501 opened the pull request:
Summary:
`batch_isend_irecv` previously only worked for two-rank cases,
otherwise it would hang, e.g. pytorch/pytorch#73960. This Diff extends
`batch_isend_irecv` to support more than two ranks. The fix is by treating
the operation more like a collective rather than two-rank P2P when selecting
the communicator, since there can be more ranks participating in the batch call than "my" rank and "my" peer.

Rules:
- If `batch_isend_irecv` is the first collective call (including collectives and
  all-to-all) in the `group` given as the argument, then all ranks of the
  `group` are expected to participate in this call.
- Otherwise, if it is not the first collective call in the `group` (i.e. the
  communicator has been initialized), then batched P2P communication involving
  only subset of processes of the `group` is allowed.

Test Plan:
Added p2p_tests.py testing the following patterns:
+    sendrecv_neighbor(input, output)       # Ring like neighbor exchange
+    sendrecv_ripple(input, output)         # Exchange with growing distance (pytorch/pytorch#73960)
+    sendrecv_P2P(input, output)            # Single P2P operation
+    isendrecv_P2P(input, output)           # Single non-blocking P2P operation
+    isendrecv_P2P_batch(input, output, 0)  # batched P2P between only two ranks
+    isendrecv_P2P_batch(input, output, 1)  # batched P2P within a new group created for two ranks

Differential Revision: D35122664

rohan-varma commented on torch/csrc/distributed/c10d/ProcessGroup.cpp:
```diff
@@ -50,7 +50,8 @@ std::string opTypeToString(OpType opType) {
 
-bool isP2POp(OpType opType) {
+bool isP2POp(OpType opType, bool batchP2P /*= false*/) {
+  if (batchP2P) return false;
```
Add a comment indicating that this basically means only dist.send/isend, dist.recv/irecv are the only ops considered p2p, and batch_isend_irecv is not considered p2p and we treat it more as a collective?

kwen2501 replied:
That's right. Will do!

rohan-varma commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -996,20 +996,20 @@ void ProcessGroupNCCL::broadcastUniqueNCCLID(
   // may create multiple NCCL communicators, so we use a sequence
   // number to differentiate between them.
-  // For point-to-point operations:
+  // For single point-to-point operations:
```
just for my understanding, does single == non-batched api version?

kwen2501 replied:
Yes, that's right :)

rohan-varma commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1100,15 +1100,18 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
   ncclUniqueId ncclID;
 
+  // For batch_isend_irecv, ncclGroupStart() would be called upfront
+  bool batchP2P = ncclActiveGroupCounter_ > 0;
```
is it safe to assume ncclGroupStart() will always increment this counter?

kwen2501 replied:
This counter is maintained by ProcessGroupNCCL instead of NCCL, here: https://github.com/pytorch/pytorch/blob/5f923fed6fcb7f0161395892c0a4310c88fbb5c4/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L2231

rohan-varma reviewed (commented):
Thanks for the fix!

It looks like `TestDistBackendWithSpawn.test_batch_isend_irecv_self_nccl` is currently failing, and can we add additional unittests to test some of the behavior that you mentioned, such as other ranks calling into the current batch as you mentioned in a comment?

kwen2501 commented:
Hi @rohan-varma thanks for bringing the test to my attention. `test_batch_isend_irecv_self_nccl` was failing because the previous form it was coded in did not conform to the following rule stated in the Summary of this PR:
>If batch_isend_irecv is the first collective call (including collectives and
all-to-all) in the group given as the argument, then all ranks of the
group are expected to participate in this call.

>Otherwise, if it is not the first collective call in the group (i.e. the
communicator has been initialized), then batched P2P communication involving
only subset of processes of the group is allowed.

I added a `dist.barrier()` in front of the `batch_isend_irecv` call to fully initialize the default group -- since it is given as the `group` argument to`batch_isend_irecv` -- then the hang went away. Another solution would be to use `dist.new_group()` to create a new group that involves only the participating ranks (would be only rank 0 in the `test_batch_isend_irecv_self_nccl` test).

I also added a new test case `test_batch_isend_irecv_ring_exchange_nccl` per your suggestion. This ring-style neighbor exchange is the most common use case of the `batch_isend_irecv` API. In this case, since all ranks of the group are participating, `dist.barrier()` style pre-initialization is not needed.

Cc @mingzhe09088

kwen2501 committed:
[PyTorch Distributed] Fix batch_isend_irecv (#74701)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/74701

Closes pytorch/pytorch#73960

`batch_isend_irecv` previously only worked for two-rank cases,
otherwise it would hang, e.g. pytorch/pytorch#73960. This Diff extends
`batch_isend_irecv` to support more than two ranks. The fix is by treating
the operation more like a collective rather than two-rank P2P when selecting
the communicator, since there can be more ranks participating in the batch call than "my" rank and "my" peer.

Rules:
- If `batch_isend_irecv` is the first collective call (including collectives and
  all-to-all) in the `group` given as the argument, then all ranks of the
  `group` are expected to participate in this call.
- Otherwise, if it is not the first collective call in the `group` (i.e. the
  communicator has been initialized), then batched P2P communication involving
  only subset of processes of the `group` is allowed.

Test Plan:
Added p2p_tests.py testing the following patterns:
+    sendrecv_neighbor(input, output)       # Ring like neighbor exchange
+    sendrecv_ripple(input, output)         # Exchange with growing distance (pytorch/pytorch#73960)
+    sendrecv_P2P(input, output)            # Single P2P operation
+    isendrecv_P2P(input, output)           # Single non-blocking P2P operation
+    isendrecv_P2P_batch(input, output, 0)  # batched P2P between only two ranks
+    isendrecv_P2P_batch(input, output, 1)  # batched P2P within a new group created for two ranks

Differential Revision: D35122664

fbshipit-source-id: 788c971aedb510bf5a1a90af59ffb20e9ed7b060

mingzhe09088 reviewed (approved):
LGTM!

kwen2501 commented:
@pytorchbot Merge this

osalpekar commented:
@pytorchbot merge this please
